pkg/bucketrequest/bucketrequest.go

```diff
@@ -0,0 +1,174 @@
+package bucketrequest
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/kubernetes-sigs/container-object-storage-interface-api/apis/objectstorage.k8s.io/v1alpha1"
+	bucketclientset "github.com/kubernetes-sigs/container-object-storage-interface-api/clientset"
+	bucketcontroller "github.com/kubernetes-sigs/container-object-storage-interface-api/controller"
+	"github.com/kubernetes-sigs/container-object-storage-interface-controller/pkg/util"
+	kubeclientset "k8s.io/client-go/kubernetes"
+
+	"github.com/golang/glog"
+)
+
+type bucketRequestListener struct {
+	kubeClient   kubeclientset.Interface
+	bucketClient bucketclientset.Interface
+}
+
+func NewListener() bucketcontroller.BucketRequestListener {
+	return &bucketRequestListener{}
+}
+
+func (b *bucketRequestListener) InitializeKubeClient(k kubeclientset.Interface) {
+	b.kubeClient = k
+}
+
+func (b *bucketRequestListener) InitializeBucketClient(bc bucketclientset.Interface) {
+	b.bucketClient = bc
+}
```

Comment:
It appears, on first sight, a `bucketRequestListener` can't reasonably be used unless its `kubeClient` and `bucketClient` are set up. As such, instead of having these `Initialize*` methods, how about requiring a `kubeclientset.Interface` and `bucketclientset.Interface` to be passed to the 'constructor' (`NewListener`)? Less API, less possibility of 'doing something wrong'.

Reply:
Some of the object require the client and some may not need. Controller listener interface is built generically. https://github.com/kubernetes-sigs/container-object-storage-interface-api/blob/43539346a903848b3f17aa0f4d41cd827d0070e1/controller/interfaces.go#L16

Reply:
Remains an odd API which can easily be mis-used. All of a sudden a consumer needs to know what any method call requires to be set in order to 'sufficiently properly' initialize an instance.